flocker/node/_deploy.py

```diff
@@ -405,13 +427,22 @@ def __init__(self, hostname, volume_service, docker_client=None,
         self.network = network
         self.volume_service = volume_service
 
-    def discover_local_state(self):
+    def discover_local_state(self, cluster_state):
         """
         List all the ``Application``\ s running on this node.
 
-        :returns: A ``Deferred`` which fires with a ``NodeState``
+        :param Deployment cluster_state: The last known state of the whole
+            cluster.
+
+        :returns: A ``Deferred`` which fires with a ``NodeApplications``
             instance.
         """
+        # XXX delete manifestations code that is unrelated to
+        # applications, return NodeApplications instance.  Application's
+        # attached volumes/manifestations can be calculated based on
+        # passed-in cluster_state. This has a potential race condition
+        # unfortunately, but probably not an issue in practice.
```

Comment:
> The failure mode in this race condition is the container agent says "hm, these apps should be using manifestations as their volumes but they aren't... and there's no manifestations!" So what does it do? Presumably it stops them.

Yep, got that.  There's a second consequence of this failure mode, though: the containers are reported back to the control service as the node's container state as running and as having no attached volume.  This is then possibly exposed to end-users who query the API for information about containers.  This will be quickly corrected (as soon as the agent stops the containers and re-reports its state) but it leaves a window open where the API is basically giving back incorrect results.

> If we're in situation where cluster state is missing info about the local node that means the agent was disconnected from the control service, perhaps for a while... a situation in which stopping the applications (which will then soon be restarted) doesn't seem like it's end of the world.

The other circumstance where it occurs is the one where the agent has just started up.  Considering the extra user-facing consequence of the race condition that I described above, this means that maybe users get randomly corrupt container information any time a container agent gets restarted.